air/src/utils.rs

```diff
@@ -0,0 +1,224 @@
+///
+/// Additionally verifies at debug time that all inputs are boolean.
+#[inline(always)]
+pub fn checked_xor<F: Field, const N: usize>(xs: [F; N]) -> F {
+    xs.into_iter().fold(F::ZERO, |acc, x| {
+        debug_assert!(x.is_zero() || x.is_one());
+        xor(acc, x)
+    })
+}
+
+/// Compute the `andnot` on a pair of field elements using the arithmetic generalization.
+///
+/// Additionally verifies at debug time that both inputs are boolean.
+#[inline(always)]
+pub fn checked_andn<F: Field>(x: F, y: F) -> F {
+    debug_assert!(x.is_zero() || x.is_one());
+    debug_assert!(y.is_zero() || y.is_one());
+    andn(x, y)
+}
+
+/// Convert a 32-bit integer into an array of 32 0 or 1 field elements.
+///
+/// The output array is in little-endian order.
+#[inline]
+pub fn u32_to_bits_le<AF: AbstractField>(val: u32) -> [AF; 32] {
+    // We do this over F::from_canonical_u32 as from_canonical_u32 can be slow
+    // like in the case of monty field.
+    array::from_fn(|i| {
+        if val & (1 << i) != 0 {
+            AF::ONE
+        } else {
+            AF::ZERO
+        }
+    })
+}
+
+/// Verify that `a = b + c + d mod 2^32`
+///
+/// We assume that a, b, c, d are all given as `2, 16` bit limbs (e.g. `a = a[0] + 2^16 a[1]`) and
+/// each `16` bit limb has been range checked to ensure it contains a value in `[0, 2^16)`.
+#[inline]
+pub fn triple_add<AB: AirBuilder>(
+    builder: &mut AB,
+    a: &[<AB as AirBuilder>::Var; 2],
+    b: &[<AB as AirBuilder>::Var; 2],
+    c: &[<AB as AirBuilder>::Expr; 2],
+    d: &[<AB as AirBuilder>::Expr; 2],
+) {
+    // Define:
+    //  acc    = a - b - c - d (mod P)
+    //  acc_16 = a[0] - b[0] - c[0] - d[0] (mod P)
+    //
+    // We perform 2 checks:
+    //
+    // (1) acc*(acc + 2^32)*(acc + 2*2^32) = 0.
+    // (2) acc_16*(acc_16 + 2^16)*(acc_16 + 2*2^16) = 0.
+    //
+    // We give a short proof for why this lets us conclude that a = b + c + d mod 2^32:
+    //
+    // As all 16 bit limbs have been range checked, we know that a, b, c, d lie in [0, 2^32) and hence
+    // a = b + c + d mod 2^32 if and only if, over the integers, a - b - c - d = 0, -2^32 or -2*2^32.
+    //
+    // Equation (1) verifies that a - b - c - d mod P = 0, -2^32 or -2*2^32.
+    //
+    // Similarly, as overflow cannot occur when computing acc_16, equation (2) verifies that
+    // over the integers, a[0] - b[0] - c[0] - d[0] = 0, -2^16 or -2*2^16. Either way
+    // we can immediately conclude that a - b - c - d = 0 mod 2^16.
+    //
+    // Now we can use the chinese remainder theorem to combine these results to conclude that
+    // a - b - c - d mod 2^16P = 0, -2^32 or -2*2^32.
+    // No overflow can occur mod 2^16 P as 2^16 P ~ 2^47 and a, b, c, d < 2^32. Hence we conclude that
+    // over the integers a - b - c - d = 0, -2^32 or -2*2^32 which implies a = b + c + d mod 2^32.
+
+    // TODO: Ideally two_16, two_32 should be saved as constants. Or at the very least
```

Comment:
Maybe `from_wrapped_u32(1 << 16)` and then `square` on that? Should be relatively cheap in case the compiler doesn't do constant propagation.

Reply:
Even better we can use `from_canonical_u32` as the function necessarily assumes that the characteristic is larger than `2^16`.